### ISSUE TITLE
xconfig for gated linear layer

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#1533]
This paper https://arxiv.org/abs/1612.08083
shows very nice results for language modeling with what they call gated convolution.  It would be very easy to implement their method in Kaldi, as a kind of variant of TDNNs.  (Since they are doing convolution in time, only, it's effectively the same thing as TDNNs).

@kangshiyin, I don't know if you have time to look into this?

Their basic idea can be represented as a 2-to-1 dimension-reducing nonlinearity f(a, b) = a * sigmoid(b), although they don't exactly implement it that way.

For prototyping purposes, it could be done as an xconfig layer that was a copy-and-modify of BasicLayer (or it might actually be easier to copy-and-modify LstmLayer).
Let's call this gated-linear-layer.
Suppose someone asked for a gated-linear-layer with output dimension 'dim', say dim=512.
Inside the layer we would first create a temporary output of dimension 2 * dim, let's call this `${layer_name}-affine`. 
It would split this via dim-range nodes into two pieces, e.g. would create config lines like:
```
dim-range-node name=${layer_name}-affine-linear input=${layer_name}-affine offset=0 dim=512 
dim-range-node name=${layer_name}-affine-presigmoid input=${layer_name}-affine offset=512 dim=512 
```
and then the presigmoid part would be fed into a SigmoidComponent, then the two pieces would be appended together as input to an ElementwiseProductComponent (e.g. with input-dim=1024 and output-dim=512, in this example).  something like...
```
component name=${layer_name} type=ElementwiseProductComponent input-dim=1024 output-dim=512
component-node name=${layer_name} component=${layer_name}  input=Append(${layer_name}-affine-linear, ${layer_name}-affine-sigmoid)
```
If this turned out to be something useful, we'd probably create a SigmoidTimesLinearComponent which combined the sigmoid and elementwise-product components into one, which would save memory and probably time also.

Note: the interaction with TDNNs here is pretty orthogonal, when you create an xconfig layer you specify a descriptor for the input which can be something like Append(-2,0,2) which expands to, say, Append(Offset(layer1, t, -2), layer1 Offset(layer1, t, 2)).  Anyway, the fact that the input is appended over time does not require any changes in the code that creates the xconfig, the input is treated as a string that represents a descriptor.  Once you start copy-and-modifying some existing code (e.g. BasicLayer or LstmLayer), you'll see.  If you modify from BasicLayer, the options that the layer accepts will have to modified less than if you modify from LstmLayer; however, you'll have to remove the code from BasicLayer that parses the name of the layer itself, and transform it into an assert, since in this GatedLinearLayer there will be only one possible layer-name ("gated-linear-layer"); we won't have options such as "relu-renorm-layer" vs. "sigmoid-layer" and so on.

This code is in steps/libs/nnet3/xconfig/.

**Comment by kangshiyin** [comment]
Sorry I probably will be busy in the following week.  Maybe @GaofengCheng will be able to help on the xconfig layer? And I can work on the CUDA version of `a * sigmoid(b)` later.

**Comment by danpovey** [comment]
[update: we'll probably need a NormalizeLayer after this.  But we might want to experiment with alternatives based on batch-norm.  So make it so the name is sigmoidgated-renorm-layer, and later we'll extend the code, like for BasicLayer, so it can be e.g. sigmoidgated-batchnorm-layer.

**Comment by danpovey** [comment]
I think @GaofengCheng is quite busy too.  This is not urgent; we'll wait for someone to pick it up.

**Comment by GaofengCheng** [comment]
@danpovey take it

**Comment by GaofengCheng** [comment]
@danpovey 
I'm running gated-tdnn and batch-norm now on AMI IHM, from log-prob, both techs are useful (results will be updated later ). I also try adding the gated-tdnn to about 25 layers, but the training will turn very bad. Also `compute_prob_train` and `compute_prob_valid` will corrupt with:
```
nnet3-chain-copy-egs --left-context=10 --right-context=10 ark:exp/ihm/chain_cleaned/tdnnsigmoidgated_1d_batchnorm_c_sp_bi/egs/train_diagnostic.cegs ark:- 
ASSERTION_FAILED (nnet3-chain-compute-prob[5.1]:AlphaGeneralFrame():chain-denominator.cc:152) : 'this_tot_alpha - this_tot_alpha == 0' 

[ Stack-Trace: ]
nnet3-chain-compute-prob() [0x116d9c2]
kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::KaldiAssertFailure_(char const*, char const*, int, char const*)
kaldi::chain::DenominatorComputation::AlphaGeneralFrame(int)
kaldi::chain::DenominatorComputation::Forward()
kaldi::chain::ComputeChainObjfAndDeriv(kaldi::chain::ChainTrainingOptions const&, kaldi::chain::DenominatorGraph const&, kaldi::chain::Supervision const&, kaldi::CuMatrixBase<float> const&, float*, float*, float*, kaldi::CuMatrixBase<float>*, kaldi::CuMatrixBase<float>*)
kaldi::nnet3::NnetChainComputeProb::ProcessOutputs(kaldi::nnet3::NnetChainExample const&, kaldi::nnet3::NnetComputer*)
kaldi::nnet3::NnetChainComputeProb::Compute(kaldi::nnet3::NnetChainExample const&)
main
__libc_start_main
nnet3-chain-compute-prob() [0xcd1bb9]
```
Any hyper-parameter need to be tuned if I want to train a very deep nnet under Nnet3?


@vijayaditya any good idea to add `res-connect` under Xconfig?
my config line like:
```
  res-connect name=res1 input=(gated_tdnn6, gated_tdnn9 )dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn10 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn11 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn12 dim=450
  res-connect name=res2 input=(res1, gated_tdnn12) dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn13 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn14 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn15 dim=450
  res-connect name=res3 input=(res2, gated_tdnn15) dim=450
```
I'm not clear about the Xconfig mechanic to process input form like` input=(res2, gated_tdnn15) `

**Comment by GaofengCheng** [comment]
@danpovey @kangshiyin @vijayaditya 
results on AMI about sigmoidgated-tdnn and batchnorm
```
System            tdnn1d_sp_bi tdnnsigmoidgated_1d_sp_bi tdnnsigmoidgated_1d_batchnorm_sp_bi
WER on dev        22.0      22.3      21.9
WER on eval        22.1      22.8      22.2
Final train prob       -0.08273-0.0652011-0.0603621
Final valid prob      -0.117588 -0.117983 -0.115286
Final train prob (xent)      -1.42143  -1.30225  -1.22445
Final valid prob (xent)      -1.56928  -1.55047  -1.49315
TDNN:
exp/ihm/chain_cleaned/tdnn1d_sp_bi: num-iters=89 nj=2..12 num-params=7.5M dim=40+100->3765 combine=-0.102->-0.099 xent:train/valid[58,88,final]=(-1.56,-1.43,-1.42/-1.64,-1.57,-1.57) logprob:train/valid[58,88,final]=(-0.099,-0.084,-0.083/-0.123,-0.118,-0.118)

sigmoidgated-TDNN+renorm:
exp/ihm/chain_cleaned/tdnnsigmoidgated_1d_sp_bi: num-iters=89 nj=2..12 num-params=11.7M dim=40+100->3765 combine=-0.087->-0.083 xent:train/valid[58,88,final]=(-1.46,-1.31,-1.30/-1.61,-1.55,-1.55) logprob:train/valid[58,88,final]=(-0.085,-0.066,-0.065/-0.119,-0.118,-0.118)

sigmoidgated-TDNN+batchnorm
exp/ihm/chain_cleaned/tdnnsigmoidgated_1d_batchnorm_sp_bi: num-iters=89 nj=2..12 num-params=11.7M dim=40+100->3765 combine=-0.084->-0.077 xent:train/valid[58,88,final]=(-1.37,-1.23,-1.22/-1.54,-1.50,-1.49) logprob:train/valid[58,88,final]=(-0.079,-0.061,-0.060/-0.117,-0.116,-0.115)

```
it seems sigmoid-gated over-fit, @danpovey  but the model capacity is stronger than traditional TDNN, I plan to try sigmoidgated-tdnn+lstm+dropout and see results, any idea?

**Comment by vijayaditya** [comment]
Could you send across the output of generate_plots.py for these three
experiments.

Vijay

**Comment by GaofengCheng** [comment]
more results:
```
System               tdnn1d_sp_bi tdnnsigmoidgated_1d_sp_bi tdnnsigmoidgated_1d_batchnorm_sp_bi tdnn_lstm1i_sp_bi_ld5 tdnn_lstm1i_perframe_dropout_sp_bi_ld5
WER on dev        22.0      22.3      21.9      20.6      19.8
WER on eval        22.1      22.8      22.2      20.1      19.2
Final train prob       -0.08273-0.0652011-0.0603621 -0.044763-0.0666221
Final valid prob      -0.117588 -0.117983 -0.115286-0.0981107 -0.097616
Final train prob (xent)      -1.42143  -1.30225  -1.22445 -0.722765 -0.915559
Final valid prob (xent)      -1.56928  -1.55047  -1.49315  -1.03985  -1.09907

```

**Comment by danpovey** [comment]
Interesting.
Remember that batch-norm is applicable to TDNNs with ReLUs as well as to the sigmoid-gated units.  @tomkocse has already done some experiments about this and it does seem to be performing a bit better than regular TDNNs with NormalizeLayer ("renorm").
It could be that the sigmoid-gated units just aren't that helpful for this type of task, but it's nice that you seem to be getting good results from per-frame dropout with regular TDNNs.

If you or @tomkocse have any improved variants of the TDNN system that you want to add to this branch-- and this would be nice-- then please make a PR to the batch_norm_and_conv branch on my personal GitHub repo.

Regarding the assertion failure-- that was a NaN-- I suspect you just had a deep nnet without using BatchNormLayers or NormalizeLayers; you need to use those.  If you were using them, let me know.

Regarding the mechanism to add residual connections, it's the following; and I'm afraid it's not very ideal:

```
 sigmoidgated-batchnorm-layer name=gated_tdnn10 input=Sum(gated_tdnn9, gated_tdnn6, gated_tdnn3) dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn11 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn12 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn13 input=Sum(gated_tdnn12, gated_tdnn9, gated_tdnn6, gated_tdnn3) dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn14 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn15 dim=450
```
It would be possible to add a 'no-op-layer' which would just produce a NoOpComponent, and which could be used to make the config like:
```
  no-op-layer input=Sum(gated_tdnn9, gated_tdnn6)
 sigmoidgated-batchnorm-layer name=gated_tdnn10 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn11 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn12 dim=450
  no-op-layer input=Sum(gated_tdnn12, gated_tdnn9)
  sigmoidgated-batchnorm-layer name=gated_tdnn13 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn14 dim=450
  sigmoidgated-batchnorm-layer name=gated_tdnn15 dim=450
```
If it turns out residual connections are helpful, and we actually make use of quite deep nnets, we could consider adding the no-op-layer (or if you make one in any case, I would merge it)..  However, I am biased against making very deep or large nnets unless there is a substantial performance improvement.

**Comment by GaofengCheng** [comment]
@danpovey 
using less epoch (2 instead of 4), gated+batch show improvement.
I'm now reducing the gated-tdnn dim to match tdnn parameter number, also using batch norm in tdnn
to figure out whether the parameter or batchnorm or gated gives the improvement.
```
System               tdnn1d_sp_bi tdnnsigmoidgated_1d_batchnorm_2epoch_sp_bi tdnnsigmoidgated_1d_batchnorm_sp_bi
WER on dev        22.0      21.4      21.9
WER on eval        22.1      21.5      22.2
Final train prob       -0.08273-0.0859849-0.0603621
Final valid prob      -0.117588 -0.118432 -0.115286
Final train prob (xent)      -1.42143  -1.45949  -1.22445
Final valid prob (xent)      -1.56928  -1.59131  -1.49315
```

**Comment by danpovey** [comment]
IIRC we didn't get any benefit from gated linear layer and probably won't merge any changes.  Closing this issue.

**Comment by danpovey** [comment]
Gaofeng, I notice this conversation stopped before there was a very clear
conclusion about whether gated linear layers were working.  What happened?

**Comment by GaofengCheng** [comment]
After:
```
System               tdnn1d_sp_bi tdnnsigmoidgated_1d_batchnorm_2epoch_sp_bi tdnnsigmoidgated_1d_batchnorm_sp_bi
parameter           7.4M    11.7M   11.7M
WER on dev        22.0      21.4      21.9
WER on eval        22.1      21.5      22.2
Final train prob       -0.08273-0.0859849-0.0603621
Final valid prob      -0.117588 -0.118432 -0.115286
Final train prob (xent)      -1.42143  -1.45949  -1.22445
Final valid prob (xent)      -1.56928  -1.59131  -1.49315
```

I have tried reduced the cell dim of `tdnnsigmoidgated_1d_batchnorm{_2epoch}_sp_bi ` from 450 to 256.
The results :
```
System               tdnn1d_sp_bitdnnsigmoidgated_1d_batchnorm_3epoch_smalldim_sp_bitdnnsigmoidgated_1d_batchnorm_3epoch_smalldim_sp_bi
parameter           7.5M    4.7M     4.7M
WER on dev        22.0      21.9      21.9
WER on eval        22.1      22.0      22.0
Final train prob       -0.08273 -0.092034 -0.092034
Final valid prob      -0.117588 -0.120451 -0.120451
Final train prob (xent)      -1.42143   -1.5237   -1.5237
Final valid prob (xent)      -1.56928  -1.64178  -1.64178
```
I observed no better than TDNN, so I gave it up after several rounds of exps.

**Comment by danpovey** [comment]
Hm.  That's interesting.  Keep the code around anyway, there is something
you might be able to try soon.
I am adding code that can decompose affine components with SVD, producing
two skinny/tall factors.
This configuration is amenable to SVD decomposition because the matrices
are closer to being square (when
we take into account the splicing).  So it might work well with that.  The
idea is we can train for some time, do
SVD, and then train some more.

**Comment by GaofengCheng** [comment]
Very interesting ! Very interesting to have a try!
